### DISCUSSION TITLE
Run django-tenants with asgi?

### DISCUSSION THREAD

**Discussion by aayushbasnet012** [django-tenants/django-tenants#1102] [Q&A]
Has anybody tried to run the django-tenants with sub-folder using uvicorn or any any async other async worker with gunicorn.
Every time i try to run with asgi, I get, Fake tenant has no attribute, sub-folder.
A little guidance in the right direction would be really helpful.

**Comment by jrutila** [comment]
I encountered this same issue while running with Daphne. I have debugged it and it seems there is some kind of asynchronous issue.

The error stems from `urlresolvers.py`. When it is getting the `tenant_prefix`, it tries to read the `schema_name` and `tenant` from `connection`. Now, the middleware in `subfolder.py` is responsible for setting that `schema_name` and `connection.tenant`. The default value being that `FakeTenant` (without `domain_subfolder`). As I printed out the `connection.tenant` it is first the fake one, but at some point it changes to the actual correct tenant.

So, could it be that, in normal WSGI run, the middleware does its thing and only then the url resolving kicks in. But in ASGI it (asynchronously?) starts the url resolving even if the middlewares are not ready yet.
